File: Strings/5/author/longSubOpt.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = dict()
        left = 0
        longest = 0
        for right in range(len(s)):
            current_char = s[right]
            if current_char in seen and seen[current_char] >= left:
                left = seen[current_char] + 1
            longest = max(longest, right-left+1)
            seen[current_char] = right
        return longest

File: Strings/5/replicate_author/longSubOpt.js
var lengthOfLongestSubstring = function (s) {
  if (s.length <= 1) return s.length;
  let maxlen = 0;
  const n = s.length;
  let left = 0;
  seenChars = {};
  for (let right = 0; right < n; right++) {
    const currentChar = s[right];
    const prevSeenChar = seenChars[currentChar]; // either undefined or the index
    if (prevSeenChar >= left) {
      left = prevSeenChar + 1;
    }
    maxlen = Math.max(maxlen, right - left + 1);
    seenChars[currentChar] = right;
  }
  return maxlen;
};

const s = 'abcbad';
console.log(lengthOfLongestSubstring(s));
